### PR TITLE
[Merged by Bors] - can define a value from inside a shader

### LINKED ISSUE #7494: Support `#define`  in shader preprocessor
## What problem does this solve or what need does it fill?

if we have 3 shader files `common.wgsl`, `util1.wgsl`, `util2.wgsl`, and if both `util1.wgsl` and `util2.wgsl` includes `common.wgsl`, then if the main shader file include both `util1.wgsl` and `util2.wgsl`, it will cause a problem, because function re-definition (i.e. `common.wgsl` imported twice).

## What solution would you like?

one solution would be to support `#define` in the shader preprocess (i.e. define a label used for `#ifdef` `#ifndef`). Then we can write an include guard similar to the include file in C.

## What alternative(s) have you considered?

there could be several other method,
1. have `#import once` declarative indicate this file should only be imported once.
2. bevy preprocessor automatically remove the duplicate function (however, how to define duplicate function is questionable)
3. do nothing, and let user keep track of import and make sure no duplicate imports happen.

### PR DESCRIPTION
**Pull request by mockersf**
# Objective

- Fixes #7494
- It is now possible to define a ShaderDef from inside a shader. This can be useful to centralise a value, or making sure an import is only interpreted once

## Solution

- Support `#define <SHADERDEF_NAME> <optional value>`

### CODE REVIEW

**Comment by devil-ira** [comment]
How does this interact with ShaderDefs defined outside the shader?
Could we (ab)use this to set default values for ShaderDefs?

**Comment by mockersf** [comment]
It overwrites them which seems the right behaviour to me

you can do something like
```
#ifndef MY_SHADER_DEF
  #define MY_SHADER_DEF 5
#endif
```

**Comment by devil-ira** [comment]
Alright, Makes sense to me

**Comment by mockersf** [comment]
bors try

File: crates/bevy_render/src/render_resource/shader.rs [hunk 1]
```diff
@@ -449,24 +456,34 @@ impl ShaderProcessor {
...
+        let mut shader_defs_unique =
+            HashMap::<String, ShaderDefVal>::from_iter(shader_defs.iter().map(|v| match v {
+                ShaderDefVal::Bool(k, _) | ShaderDefVal::Int(k, _) | ShaderDefVal::UInt(k, _) => {
+                    (k.clone(), v.clone())
+                }
+            }));
+        self.process_inner(shader, &mut shader_defs_unique, shaders, import_handles)
+    }
+
+    fn process_inner(
```
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/shader.rs:469]: I'm not sure I understand why this needed to be split in process/process_inner

Comment (IceSentry) [reply crates/bevy_render/src/render_resource/shader.rs:469]: Ah, it's because it's used in apply_imports

**Review by IceSentry — APPROVED** [review]
LGTM, the functionality is really nice.

Side note: `ShaderProcessor::process_inner()` is getting pretty big, I wonder if there's a way to split it up a bit.

File: crates/bevy_render/src/render_resource/shader.rs (hunk 1 above)
Comment (mockersf) [reply crates/bevy_render/src/render_resource/shader.rs:469]: Processing a shader need to be able to call itself recursively, and now to modify the shader defs while a given shader is processed. But values defined in one shader should not impact another, so the shader defs modifications are not bubbled up to the outside caller of `process`

File: crates/bevy_render/src/render_resource/shader.rs [hunk 2]
```diff
@@ -544,6 +561,26 @@ impl ShaderProcessor {
...
+                let def = cap.get(1).unwrap();
+                let name = def.as_str().to_string();
+
+                if let Some(val) = cap.get(2) {
+                    if let Ok(val) = val.as_str().parse() {
+                        shader_defs_unique.insert(name.clone(), ShaderDefVal::UInt(name, val));
+                    } else if let Ok(val) = val.as_str().parse() {
+                        shader_defs_unique.insert(name.clone(), ShaderDefVal::Int(name, val));
+                    } else if let Ok(val) = val.as_str().parse() {
+                        shader_defs_unique.insert(name.clone(), ShaderDefVal::Bool(name, val));
```
Comment (superdump) [inline crates/bevy_render/src/render_resource/shader.rs:574]: The type inference here makes for confusing-looking code. Could you use explicit types in the `.parse::<T>()` to make it clearer?

Comment (superdump) [reply crates/bevy_render/src/render_resource/shader.rs:574]: Or... could we go C-preprocessor-style and just use string replacement? When I got down to the tests with `DEFUINT 1` and `DEFINT -1` it felt weird to define the type of the value here given that `1` in WGSL is signed and `1u` is unsigned.

Comment (mockersf) [reply crates/bevy_render/src/render_resource/shader.rs:574]: > The type inference here makes for confusing-looking code. Could you use explicit types in the `.parse::<T>()` to make it clearer?

but I like the type magic that makes the same code not have the same result 😄 

> Or... could we go C-preprocessor-style and just use string replacement? When I got down to the tests with `DEFUINT 1` and `DEFINT -1` it felt weird to define the type of the value here given that `1` in WGSL is signed and `1u` is unsigned.

keeping it as string all along would make harder to use the value in if condition

**Review by superdump — APPROVED** [review]

**Pull request closed without merging**